### PR TITLE
Add support for RISC-V 32bit / RISC-V 64bit / LoongArch 64-bit

### PR DESCRIPTION
**Pull request by bcoles**
This PR adds basic support for RISCV and LoongArch by defining the following constants for use within Metasploit Framework and accompanying `rex-*` libraries:

* `RISCV32BE`
* `RISCV32LE`
* `RISCV64BE`
* `RISCV64LE`
* `LOONGARCH64`

# RISC-V

RISC-V is rapidly gaining popularity. The first three [mentions of RISC-V on the Metasploit issue tracker](https://github.com/rapid7/metasploit-framework/issues?q=risc-v) occurred within the past 6 months.

# LoongArch

This PR adds support for LoongArch 64-bit - not 32-bit.

`LOONGARCH64` was chosen as the constant name (rather than `LOONG64`) to differentiate from [Loongson](https://en.wikipedia.org/wiki/Loongson) chips and 32-bit predecessors.

The [Linux kernel documentation](https://lwn.net/Articles/861951/) offers the following:

> LoongArch is a new RISC ISA, which is a bit like MIPS or RISC-V.
> LoongArch includes a reduced 32-bit version (LA32R), a standard 32-bit
> version (LA32S) and a 64-bit version (LA64).

Candidates for future constant names are:

* `LOONGARCH32`
* `LOONGARCH32R`
* `LOONGARCH32S`

I have not been able to emulate a LoongArch 32-bit system (due to limited community interest and limited personal interest).

### CODE REVIEW

**Review by jmartin-tech — COMMENTED** [review]
I suspect this will meet initial use cases.

Is a single entry each for `riscv32` and `riscv64` sufficient? I *think* the answer is `yes`.

Look like the [Linux Kernel](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/riscv/Makefile?h=v6.11&id=516fca5a7516cde7a9968f84179ed20ffb438885) is reporting `riscv64` and `riscv32` to `proc`. If I am reading the `Makefile` correctly any `rv32ima` and `rv64ima` combination will be rolled up as `riscv##` when exposed in proc.

The [toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain) however offers a number of sub architectures that might be used at compile time for a binary:

```
--with-arch=rv64imafdc  Sets the base RISC-V ISA, defaults to rv64imafdc
--with-abi=lp64d        Sets the base RISC-V ABI, defaults to lp64d
--with-tune=rocket      Set the base RISC-V CPU, defaults to rocket
 
glibc_multilib_names="rv32imac-ilp32 rv32imafdc-ilp32d rv64imac-lp64 rv64imafdc-lp64d"
 
glibc_multilib_names="rv32i-ilp32 rv32iac-ilp32 rv32im-ilp32 rv32imac-ilp32 rv32imafc-ilp32f rv64imac-lp64 rv64imafdc-lp64d"
 
musl_multilib_names="rv64imac-lp64 rv64imafdc-lp64d"
```

see also:
https://github.com/riscv-collab/riscv-gnu-toolchain/blob/a504f73f40d556aaa9a55b3d4377abba0176c2b5/regression/Makefile#L6-L31

**Comment by bcoles** [comment]
> Look like the [Linux Kernel](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/riscv/Makefile?h=v6.11&id=516fca5a7516cde7a9968f84179ed20ffb438885) is reporting `riscv64` and `riscv32` to `proc`. If I am reading the `Makefile` correctly any `rv32ima` and `rv64ima` combination will be rolled up as `riscv##` when exposed in proc.

This matches my experience with the few boards and emulated systems I've used.

I have the following complementary change to Metasploit `Linux::Post` library (following update of `rex-arch`) which uses the kernel's self-reported architecture from `uname -m`.

```diff
diff --git a/lib/msf/core/post/linux/kernel.rb b/lib/msf/core/post/linux/kernel.rb
index ae1191f115..6dca23718a 100644
--- a/lib/msf/core/post/linux/kernel.rb
+++ b/lib/msf/core/post/linux/kernel.rb
@@ -71,6 +71,9 @@ module Kernel
     return ARCH_MIPS if arch == 'mips'
     return ARCH_MIPS64 if arch == 'mips64'
     return ARCH_SPARC if arch == 'sparc'
+    return ARCH_RISCV32 if arch == 'riscv32'
+    return ARCH_RISCV64 if arch == 'riscv64'
+    return ARCH_LOONGARCH64 if arch == 'loongarch64'
     arch
   end
 ```

**Comment by bcoles** [comment]
> The [toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain) however offers a number of sub architectures that might be used at compile time for a binary:
> 
> ```
> --with-arch=rv64imafdc  Sets the base RISC-V ISA, defaults to rv64imafdc
> --with-abi=lp64d        Sets the base RISC-V ABI, defaults to lp64d
> --with-tune=rocket      Set the base RISC-V CPU, defaults to rocket
>  
> glibc_multilib_names="rv32imac-ilp32 rv32imafdc-ilp32d rv64imac-lp64 rv64imafdc-lp64d"
>  
> glibc_multilib_names="rv32i-ilp32 rv32iac-ilp32 rv32im-ilp32 rv32imac-ilp32 rv32imafc-ilp32f rv64imac-lp64 rv64imafdc-lp64d"
>  
> musl_multilib_names="rv64imac-lp64 rv64imafdc-lp64d"
> ```
> 
> see also: https://github.com/riscv-collab/riscv-gnu-toolchain/blob/a504f73f40d556aaa9a55b3d4377abba0176c2b5/regression/Makefile#L6-L31

I don't think it's worth differentiating between these, in the same way that we don't differentiate between ARM and ARM with Thumb extensions, and ARMv6/7/8/9....

However, it wouldn't hurt to be explicit in terms of endianess, as RISC-V is bi-endian (like ARM). We can specify BE/LE in the constant name.

**Comment by bcoles** [comment]
> However, it wouldn't hurt to be explicit in terms of endianess, as RISC-V is bi-endian (like ARM). We can specify BE/LE in the constant name.

I've force pushed this change to differentiate between little endian and big endian.

**Review by adfoster-r7 — APPROVED** [review]
Looks good to me; but will wait for @jmartin-tech's final review before merging

**Review by jmartin-tech — APPROVED** [review]

**Pull request merged**
